foundationdb/src/transaction.rs

```diff
@@ -287,6 +509,167 @@ impl Future for TrxGet {
+                trx: self.trx.clone(),
+                inner: r,
+            })),
+            Ok(Async::NotReady) => Ok(Async::NotReady),
+            Err(e) => Err(e),
+        }
+    }
+}
+
+/// Represents the data of a `Transaction::get_range`
+pub struct GetRangeResult {
+    trx: Transaction,
+    inner: FdbFutureResult,
+}
+impl GetRangeResult {
+    /// Returns a clone of the Transaction this get is a part of
+    pub fn transaction(&self) -> Transaction {
+        self.trx.clone()
+    }
+
+    /// Returns the values associated with this get
+    pub fn keyvalues(&self) -> Result<KeyValues> {
+        self.inner.get_keyvalue_array()
+    }
+}
+
+/// A future results of a `get_range` operation
+struct TrxGetRange {
+    trx: Transaction,
+    inner: FdbFuture,
+}
+impl Future for TrxGetRange {
+    type Item = GetRangeResult;
+    type Error = FdbError;
+
+    fn poll(&mut self) -> std::result::Result<Async<Self::Item>, Self::Error> {
+        match self.inner.poll() {
+            Ok(Async::Ready(r)) => Ok(Async::Ready(GetRangeResult {
+                trx: self.trx.clone(),
+                inner: r,
+            })),
+            Ok(Async::NotReady) => Ok(Async::NotReady),
+            Err(e) => Err(e),
+        }
+    }
+}
+
+/// TODO
+pub struct KeyValuesStream {
+    //TODO: use `RangeOption` instead?
+    begin: OwnedKeySelector,
+    end: OwnedKeySelector,
+
+    limit: usize,
+    target_bytes: usize,
+    mode: options::StreamingMode,
+    iteration: usize,
+    snapshot: bool,
+    reverse: bool,
+
+    trx: Transaction,
+    inner: Option<TrxGetRange>,
+    index: usize,
+}
+
+impl KeyValuesStream {
+    fn update_inner(&mut self) {
+        self.iteration += 1;
+        self.inner = Some(self.trx.get_range_inner(
+            self.begin.as_selector(),
+            self.end.as_selector(),
+            self.limit - self.index,
+            self.target_bytes,
+            self.mode,
+            self.iteration,
+            self.snapshot,
+            self.reverse,
+        ));
+    }
+
+    fn advance(&mut self, res: &GetRangeResult) -> Result<()> {
+        let kvs = res.keyvalues()?;
+        let more = kvs.more();
+
+        if !more {
+            return Ok(());
+        }
+
+        let slice = kvs.as_ref();
+        if slice.is_empty() {
+            return Ok(());
+        }
+        self.index += slice.len();
+
+        let last = slice.last().unwrap();
+        let last_key = last.key();
+
+        if self.reverse {
+            self.end = KeySelector::first_greater_or_equal(last_key).to_owned();
+        } else {
+            self.begin = KeySelector::first_greater_than(last_key).to_owned();
+        }
+
+        self.update_inner();
+        Ok(())
+    }
+}
+
+impl<'a> Stream for KeyValuesStream {
+    type Item = GetRangeResult;
+    type Error = FdbError;
+
+    fn poll(&mut self) -> std::result::Result<Async<Option<Self::Item>>, Self::Error> {
+        if self.inner.is_none() {
```

Comment:
Would it be here that we'd track the `more` flag and then trigger another query to get results after the last key received?